include/utils.hpp

```diff
@@ -126,6 +137,28 @@ precondition(bool cond) {
 }
 #endif
 
+template<auto X>
+inline constexpr std::integral_constant<std::remove_const_t<decltype(X)>, X> ic{};
```

Comment:
I do not like this definition because the `ic` shortened definition obfuscates the in-code usage. Since this isn't used much (and only in expert-level code area) I'd suggest to use this as is there...

Reply:
If this shortened definition is really necessary, then I'd recommend to declare it using `using` close to the location where it is actually used/needed. The risk of these many little helper definitions is that it widens the effective C++ standard/API surface that needs to be maintained.

Reply:
I added it because I believe it makes the code more readable. If I'm alone with that opinion I'm fine with removing `ic`.